datafusion/expr/src/expr_schema.rs

```diff
@@ -68,7 +69,26 @@ impl ExprSchemable for Expr {
             Expr::OuterReferenceColumn(ty, _) => Ok(ty.clone()),
             Expr::ScalarVariable(ty, _) => Ok(ty.clone()),
             Expr::Literal(l) => Ok(l.get_datatype()),
-            Expr::Case(case) => case.when_then_expr[0].1.get_type(schema),
+            Expr::Case(case) => {
+                // when #5681 will be fixed, this code can be reverted to:
```

Comment:
How about moving TypeCoersion to `analyzer` to fix #5733 instead of making a temporary workaround?

Reply:
I agree that it's a duck-taping.

I'm afraid the mentioned issue **won't be resolved quickly** by cut'n'paste... (see https://github.com/apache/arrow-datafusion/issues/5681#issuecomment-1483780310 )

So the question is - **should we expose our users to the bug** 🧐 until #5681 is fixed? (putting Datafusion's **reputation** at risk?)

What is more, this PR features **sqlogictest** ⭐️, so we won't have a regression caused by the fix #5681.

Reply:
The type coercion is implemented as an optimizer - and it changes schemas of particular `Expr`essions.
And it **takes care** of `Expr::Case`.

But `Expr::Projection` caches the schema in itself - so it is not updated! Therefore we get a mismatch between declared schema vs batch schema.

Essentially this PR makes `Expr::Case` to **not** change a schema during the coercion optimizer.

Reply:
Filed https://github.com/apache/arrow-datafusion/issues/5821